Scope based enum

3nids opened the pull request:
This is prototype with adding scope based.
This allows documenting them properly

![image](https://user-images.githubusercontent.com/127259/53985971-43458f80-411d-11e9-97a7-34982a0756e9.png)
![image](https://user-images.githubusercontent.com/127259/53985984-4c366100-411d-11e9-92e0-391ff8dbd3c7.png)

Python API compatibility is maintained by monkey patching.

3nids committed:
[sipify] handle scope based enums

3nids committed:
use scope based enum for QgsAdvancedDigitizingDockWidget::AditionalConstraints

3nids committed:
[sipify] monkey patch enum for PyQGIS API compatibility

3nids committed:
run sipify

nyalldawson commented:
Very nice!!

What's the single "noconstraint" dot point in the top screenshot for?

3nids committed:
[sipify] fix oneliners

3nids committed:
[sipify] add macro to explicitely ask for monkey patching

because new scope based enum should not do the monkey patching, it's only done for python API compatibility

m-kuhn commented:
Should we...

move global enums to `Qgis`, so it would be `Qgis::MapLayerType`, that would be aligned with Qt, no?
